tools/tpm2_policypcr.c

```diff
@@ -0,0 +1,165 @@
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <errno.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+#include <ctype.h>
+
+#include <sapi/tpm20.h>
+
+#include "files.h"
+#include "log.h"
+#include "pcr.h"
+#include "tpm2_options.h"
+#include "tpm2_policy.h"
+#include "tpm2_session.h"
+#include "tpm2_tool.h"
+#include "tpm2_util.h"
+
+typedef struct tpm2_startauthsession_ctx tpm2_startauthsession_ctx;
+struct tpm2_startauthsession_ctx {
+   const char *session_path;
+   const char *raw_pcrs_file;
+   TPML_PCR_SELECTION pcr_selection;
+   const char *policy_out_path;
+};
+
+static tpm2_startauthsession_ctx ctx;
+
+static bool on_option(char key, char *value) {
+
+    switch (key) {
+    case 'f':
+        ctx.policy_out_path = value;
+        break;
+    case 'F':
+        ctx.raw_pcrs_file = value;
+        break;
+    case 'L': {
+        bool result = pcr_parse_selections(value, &ctx.pcr_selection);
+        if (!result) {
+            LOG_ERR("Could not parse PCR selections");
+            return false;
+        }
+    } break;
+    case 'S':
+        ctx.session_path = value;
+    break;
+    }
+    return true;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+
+    static struct option topts[] = {
+        { "policy-file",    required_argument,  NULL,   'f' },
+        { "pcr-input-file", required_argument,  NULL,   'F' },
+        { "set-list",       required_argument,  NULL,   'L' },
+        { "session",        required_argument,  NULL,   'S' },
+    };
+
+    *opts = tpm2_options_new("f:F:L:S:", ARRAY_LEN(topts), topts, on_option,
+                             NULL, false);
+
+    return *opts != NULL;
+}
+
+int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
+
+    UNUSED(flags);
+
+    int rc = 1;
+    tpm2_session *s = NULL;
+
+    bool fail = false;
+    if (!ctx.session_path) {
+        LOG_ERR("Must specify -S session file");
+        fail = true;
+    }
+
+    if (!ctx.pcr_selection.count) {
+        LOG_ERR("Must specify -L pcr selection list");
+        fail = true;
+    }
+
+    if (fail) {
+        return 1;
+    }
+
+    s = tpm2_session_restore(ctx.session_path);
+    if (!s) {
+        return false;
+    }
+
+
+    bool result = tpm2_policy_build_pcr(sapi_context, s,
+            ctx.raw_pcrs_file,
+            &ctx.pcr_selection);
+    if (!result) {
+        LOG_ERR("Could not build pcr policy");
+        goto out;
+    }
+
+    TPM2B_DIGEST policy_digest = TPM2B_EMPTY_INIT;
+    result = tpm2_policy_get_digest(sapi_context, s,
+            &policy_digest);
+    if (!result) {
+        LOG_ERR("Could not build tpm policy");
+        goto out;
+    }
+
+    tpm2_tool_output("policy_digest: 0x");
```

Comment:
Again I wonder if we shouldn't just print the policy digest and in any case print additional info with -V.

I think most users that care about this information will prefer to just do:

```
digest=`tpm2_policypcr -S session.dat -L "sha1:0,1,2,3" -F pcr.dat -f policy.dat`
```

Reply:
it doesn't have multiple returns per-the-spec, so offhand I don't see why not. I just think having keys in the output helps from a human perspective at the expense of making scripts slightly more clunky.
``
tpm2_policypcr <args>
0x123456  # <-- what is this?
``
vs
```
tpm2_policypcr <args>
policy_digest: 0x123456 # <-- Oh its a policy_digest!
```